kotlinx-coroutines-core/common/src/sync/ReadWriteMutex.kt

```diff
@@ -0,0 +1,603 @@
+ * not acquired and takes the lock immediately in this case. Otherwise, if the writer should
+ * wait for the lock, the algorithm increases the waiting writers counter and suspends.
+ *
+ * While releasing a reader lock, the algorithm decrements the number of active readers.
+ * However, if the counter reaches zero, it checks whether a writer is waiting for the lock
+ * and resumes the first waiter. The writer lock release resumes either the next waiting writer
+ * decrementing the number of them or all waiting readers decrementing the corresponding counter.
+ * When both readers and writers are waiting, a truly fair implementation should check whether
+ * the first reader or the first writer came earlier and resume it. Simultaneously, it potentially
+ * should not resume all waiting readers since it is possible for several readers to come following by
+ * a writer and several readers again. In this case, only the first group of readers should be resumed
+ * in a truly fair implementation. However, our implementation does not support this behavior and resumes
+ * either the first writer or all the waiting readers. In some sense, it can prioritize readers or writers.
+ * To make the algorithm practically fair, so that neither readers nor writers starve, we flip a coin
+ * to choose whether writers or readers should be prioritized this time.
+ *
+ * As for cancellation, the main idea is to revert the state update. However, possible logical races
+ * should be managed, which makes the revert part non-trivial. The details are discussed in the code
+ * comments and appear almost everywhere.
+ * */
+@OptIn(HazardousConcurrentApi::class)
+internal class ReadWriteMutexImpl : ReadWriteMutex {
+    // The number of coroutines waiting for a reader lock in `sqsReaders`.
+    private val waitingReaders = atomic(0)
+    // This state field contains several counters
+    // and is always updated atomically by `CAS`:
+    // - `AR` (active readers) is a 30-bit counter which represents the number
+    //                         of coroutines holding a read lock;
+    // - `WLA` (write lock acquired) is a flag which is `true` when the write lock is acquired;
+    // - `WW` (waiting writers) is a 30-bit counter which represents the number
+    //                          of coroutines waiting for the write lock in `sqsWriters`;
+    // - `RWR` (resuming waiting readers) is a flag which is `true` when waiting readers
+    //                                    resumption is in progress.
+    private val state = atomic(0L)
+
+    private val sqsReaders = ReadersSQS() // The place where readers should suspend and be resumed.
+    private val sqsWriters = WritersSQS() // The place where writers should suspend and be resumed.
+
+    override suspend fun readLock() {
+        // Try to acquire a reader lock without suspension at first,
+        // this can be considered as a performance optimization and
+        // does not affect correctness.
```

Comment:
It does. If not for this line, `readLock()` would spin until a writer arrived.

Reply:
Agree, removed this note